### PR TITLE
util/codec,util/errctx,*: add error context, and use it to refactor `util/codec`

### LINKED ISSUE #48611: Refractor the way to handle error and warning
## Enhancement

While I'm trying to refractor `rowcodec`, `tablecodec`, `codec`... package, I found it's hard because we don't have an unified way to handle the errors and warnings. We can use a `errctx` to store all the flag related with error handling (especially from `SQLMode`...).

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #48611

Problem Summary:

This PR adds an error context to handle the error. The function signature is:

```
func (ctx *Context) HandleError(err error) error

func (ctx *Context) HandleErrorWithAlias(internalErr error, err error, warnErr error) error
```

According to the `internalErr` and the error level, it can choose to return the `err` or append the `warnErr` to warnings. This PR also uses this context to modify the `util/codec` package to show its ability (and make this PR do have some effect). The main part of this PR is the first commit: https://github.com/YangKeao/tidb/commit/75401b5812032bf35d96afa94e14f1e7b31e60ce . 

Another thing may need to discuss is about how to combine multiple contexts. I used two separated arguments here. Would it be better to use a single one (e.g. `CodecContext`, which contains the type context and error context) and unifying them as one argument?

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Other refractors are covered by existing tests.

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/util/codec/BUILD.bazel [hunk 1]
```diff
@@ -12,9 +12,9 @@ go_library(
     importpath = "github.com/pingcap/tidb/pkg/util/codec",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/errctx",
         "//pkg/parser/mysql",
         "//pkg/parser/terror",
-        "//pkg/sessionctx/stmtctx",
```
Comment (YangKeao) [inline pkg/util/codec/BUILD.bazel:17]: Once this PR is merged. I can continue this process on `rowcodec`, `tablecodec`... They generally also just need `typeCtx` and `errCtx`.

**Review by hawkingrei — APPROVED** [review]

File: pkg/errctx/context.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		levelMap:        ctx.levelMap,
+		appendWarningFn: ctx.appendWarningFn,
+	}
+	newCtx.levelMap[eg] = l
+
+	return newCtx
+}
+
+// AppendWarning appends the error to warning. If the inner `appendWarningFn` is nil, do nothing.
+func (ctx *Context) AppendWarning(err error) {
```
Comment (lcwangchao) [inline pkg/errctx/context.go:62]: Is it possible to make `AppendWarning` private? I think it is not expected to be called directly by other code.

File: pkg/errctx/context.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+})
+
+var errGroupMap = make(map[errors.ErrCode]ErrGroup)
+
+// ErrGroup groups the error according to the behavior of handling errors
+type ErrGroup int
+
+const (
+	// ErrGroupTruncate is the group of truncated errors
+	ErrGroupTruncate ErrGroup = iota
```
Comment (lcwangchao) [inline pkg/errctx/context.go:129]: Will `types.FlagTruncateAsWarnning` be removed in the future. Their semantics are same

File: pkg/util/codec/codec.go [hunk 4]
```diff
@@ -293,17 +289,17 @@ func sizeInt(comparable1 bool) int {
 // EncodeKey appends the encoded values to byte slice b, returns the appended
 // slice. It guarantees the encoded value is in ascending order for comparison.
 // For decimal type, datum must set datum's length and frac.
-func EncodeKey(sc *stmtctx.StatementContext, b []byte, v ...types.Datum) ([]byte, error) {
-	return encode(sc, b, v, true)
+func EncodeKey(typeCtx types.Context, errctx errctx.Context, b []byte, v ...types.Datum) ([]byte, error) {
```
Comment (lcwangchao) [inline pkg/util/codec/codec.go:292]: How about simplifying the methods `EncodeKey` and `EncodeValue` to:

```
func EncodeKey(loc *time.Location, b[]byte, v ...types.Datum)
func EncodeValue(loc *time.Location, b []byte, v ...types.Datum)
```

And we should handle the error outside encode methods like this:

```
b, err := EcodeKey(loc, b, v);
err = ectx.HandleError(err, err, err)
```

File: pkg/errctx/context.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+// In most cases, these three should be the same. If there are many different kinds of error internally, but they are expected
+// to give the same error to users, the `err` can be different form `internalErr`. Also, if the warning is expected to be
+// different from the initial error, you can also use the `warnErr` argument.
+//
+// TODO: would it be better to use `HandleError(err error, ...error)`? As the second and third arguments are the same with
+// the first one is most cases.
+// TODO: is it good to give an error code for internal only errors? Or should we use another way to distinguish different
+// group of errors?
+// TODO: both `types.Context` and `errctx.Context` can handle truncate error now. Refractor them.
+func (ctx *Context) HandleError(internalErr error, err error, warnErr error) error {
```
Comment (lcwangchao) [inline pkg/errctx/context.go:85]: This method has three arguments but the second and the third is the same with the first one in most cases. Can we rename this method to `HandleErrorWithAlias` and then make `HandleError` with only one parameter and same with `HandleErrorWithAlias(err, err, err)`

File: pkg/errctx/context.go (hunk 3 above)
Comment (YangKeao) [reply pkg/errctx/context.go:129]: Yes. I have a TODO in the comment of `HandleError`:

```
// TODO: both `types.Context` and `errctx.Context` can handle truncate error now. Refractor them.
```

File: pkg/errctx/context.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	LevelError Level = iota
+	// LevelWarn means it will be regarded as a warning
+	LevelWarn
+	// LevelIgnore means the error will be ignored
+	LevelIgnore
+)
+
+// Context defines how to handle an error
+type Context struct {
+	levelMap        [2]Level
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/errctx/context.go:37]: I know what you mean. It's actually `errGroupCount`, right? Nice suggestion.

File: pkg/util/codec/codec.go (hunk 4 above)
Comment (YangKeao) [reply pkg/util/codec/codec.go]: Done.

File: pkg/errctx/context.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+const errGroupCount = 2
+
+// ErrGroup groups the error according to the behavior of handling errors
+type ErrGroup int
+
+const (
+	// ErrGroupTruncate is the group of truncated errors
+	ErrGroupTruncate ErrGroup = iota
+	// ErrGroupOverflow is the group of overflow errors
+	ErrGroupOverflow
```
Comment (lcwangchao) [inline pkg/errctx/context.go:140]: could we put `errGroupCount` here so that it can auto-increase when we have new groups?

**Review by lcwangchao — COMMENTED** [review]
I think we can also refactor package `util/rowcodec` like `util/code`

File: pkg/errctx/context.go (hunk 7 above)
Comment (YangKeao) [reply pkg/errctx/context.go:138]: Done.

**Comment by YangKeao** [comment]
> I think we can also refactor package `util/rowcodec` like `util/code`
> 
> Rest LGTM

I'll refractor them (`util/rowcodec` and `util/tablecodec`, etc) in the following PR.

**Review by lcwangchao — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Review by tangenta — APPROVED** [review]

**Review by Leavrth — APPROVED** [review]

**Pull request merged**